### PR TITLE
service: paxos_state: wean off get_local_storage_proxy()

### PR DESCRIPTION
**Pull request by avikivity**
Instead of calling get_local_storage_proxy in paxos_state, get it from the
caller (who is, in fact, storage_proxy or one of its components).

Some of the callers, although they are storage_proxy components, don't
have a storage_proxy reference handy and so they ignomiously call
get_local_storage_proxy() themselves. This will be adjusted later.

The other callers who are, in fact, storage_proxy, have to take special
care not to cross a shard boundary. When they do, smp::submit_to()
is converted to sharded::invoke_on() in order to get the correct local instance.

Test: unit (dev)

### CODE REVIEW

**Review by nyh — APPROVED** [review]

File: service/storage_proxy.cc
Comment (gleb-cloudius) [inline service/storage_proxy.cc:343]: You have `sp` here. Why call `get_local_storage_proxy`?

File: service/storage_proxy.cc
Comment (gleb-cloudius) [inline service/storage_proxy.cc:970]: Here you have storage_proxy pointer in this.

File: service/storage_proxy.cc
Comment (nyh) [reply service/storage_proxy.cc:343]: Good catch, I missed this in my review. @avikivity I already committed this patch, so you can fix it as a followup.

File: service/storage_proxy.cc
Comment (gleb-cloudius) [inline service/storage_proxy.cc:1131]: And here.

**Comment by gleb-cloudius** [comment]
It would be nice to wait more than 10 minutes before merging a pull request.

**Comment by nyh** [comment]
> It would be nice to wait more than 10 minutes before merging a pull request.

Maybe, though I'm not sure the people who send PRs agree with you.

Please note that none of the things you found can't be changed later. Avi himself wrote in the commit message "they ignomiously call get_local_storage_proxy() themselves. This will be adjusted later.". Avi (and I in review) indeed missed some get_local_storage_proxy() calls that could be trivially replaced now instead of "later", but he knows he'll need to look at all of them anyway.

**Comment by gleb-cloudius** [comment]
> > It would be nice to wait more than 10 minutes before merging a pull request.
> 
> Maybe, though I'm not sure the people who send PRs agree with you.
> 

Then let them just commit it straight. Some PRs are committed in10 minutes without any real chance to be reviews and some sits for weeks without been looked at.

> Please note that none of the things you found can't be changed later. Avi himself wrote in the commit message "they ignomiously call get_local_storage_proxy() themselves. This will be adjusted later.". Avi (and I in review) indeed missed some get_local_storage_proxy() calls that could be trivially replaced now instead of "later", but he knows he'll need to look at all of them anyway.

Anything can be changed later.

**Comment by nyh** [comment]
> > > It would be nice to wait more than 10 minutes before merging a pull request.
> > 
> > 
> > Maybe, though I'm not sure the people who send PRs agree with you.
> 
> Then let them just commit it straight. Some PRs are committed in10 minutes without any real chance to be reviews and some sits for weeks without been looked at.

First, I did review it. I did not carefully review all get_local_storage_proxy() calls for a chance to be replaced with something else, because Avi explicitly asked to do it later (I agree that some of the things you found were trivial to fix immediately).

Second, there is no real difference between a patch being reviewed in 10 minutes, and a patch sitting on the mailing list for a week and *then* being reviewed in 10 minutes. In both cases it got the same 10-minute review. It seems like you're hoping that each patch be carefully reviewed by 3 different people, and only after these 3 people reach a consensus, the patch gets committed. But this doesn't happen for most of the patches.
 
> > Please note that none of the things you found can't be changed later. Avi himself wrote in the commit message "they ignomiously call get_local_storage_proxy() themselves. This will be adjusted later.". Avi (and I in review) indeed missed some get_local_storage_proxy() calls that could be trivially replaced now instead of "later", but he knows he'll need to look at all of them anyway.
> 
> Anything can be changed later.

Right, which is why I think it's important to avoid getting bugs into the code, but less important to insist on perfect implementations when it can be perfected later.

**Comment by gleb-cloudius** [comment]
> > > > It would be nice to wait more than 10 minutes before merging a pull request.
> > > 
> > > 
> > > Maybe, though I'm not sure the people who send PRs agree with you.
> > 
> > 
> > Then let them just commit it straight. Some PRs are committed in10 minutes without any real chance to be reviews and some sits for weeks without been looked at.
> 
> First, I did review it. I did not carefully review all get_local_storage_proxy() calls for a chance to be replaced with something else, because Avi explicitly asked to do it later (I agree that some of the things you found were trivial to fix immediately).
> 
> Second, there is no real difference between a patch being reviewed in 10 minutes, and a patch sitting on the mailing list for a week and _then_ being reviewed in 10 minutes. In both cases it got the same 10-minute review. It seems like you're hoping that each patch be carefully reviewed by 3 different people, and only after these 3 people reach a consensus, the patch gets committed. But this doesn't happen for most of the patches.

There is a real and huge difference between patch been on the mailing list for more then 10 minutes because then it has a chance to be reviews by more people and and we will not get into a deal where maintainers make each one a favor of reviewing and pushing each other patches without anybody else get a chance to look at them.

And of course you need the code owner to review the code before you pushing it as a minimum.

> 
> > > Please note that none of the things you found can't be changed later. Avi himself wrote in the commit message "they ignomiously call get_local_storage_proxy() themselves. This will be adjusted later.". Avi (and I in review) indeed missed some get_local_storage_proxy() calls that could be trivially replaced now instead of "later", but he knows he'll need to look at all of them anyway.
> > 
> > 
> > Anything can be changed later.
> 
> Right, which is why I think it's important to avoid getting bugs into the code, but less important to insist on perfect implementations when it can be perfected later.

To have more chance to not introduce bugs you need to get an opportunity for the code to be reviewed by more people and even then they will be introduced.

**Comment by nyh** [comment]
> There is a real and huge difference between patch been on the mailing list for more then 10 minutes because then it has a chance to be reviews by more people and and we will not get into a deal where maintainers make each one a favor of reviewing and pushing each other patches without anybody else get a chance to look at them.

Take a look at most other patches. You'll sadly see that most of them (though obviously not all) sat on the mailing list or PR for anything between 10 minutes and 10 days - and then got reviewed by one person and committed by the same person. In 75% of the cases this one person was Avi. So you are suggesting a great mechanism, but it's not how we've been doing things.

> And of course you need the code owner to review the code before you pushing it as a minimum.

Maybe, although this hasn't been the "minimum" until now. 

> To have more chance to not introduce bugs you need to get an opportunity for the code to be reviewed by more people and even then they will be introduced.

There's always a tradeoff here between waiting for multiple reviews and quick commit, and it depends (I think) on the "risk" of the patch. If I don't feel comfortable in my understanding of the code, or feel that the person I'm reviewing doesn't know what he's doing - I'll usually wait for more reviewers. If I think the submitter of the code knew what he was doing and the patch doesn't have hard-to-understand algorithms and it's just some straightforward cleanup, I might think it doesn't need more reviews. This is what I did in this case.

BTW, you seem to be missing the fact that I *did* review this code. It's not like I just took random junk code sent by a first-time contributer and put it in without even looking at it. It was code from a really good developer, and reviewed by me.

And yes, clearly reviews aren't perfect. Even when they're long. That's all the more reason to invest more in other ways of making code perfect (e.g., tests) and not insist on extended reviews for even the simplest of patches.

**Comment by gleb-cloudius** [comment]
> > There is a real and huge difference between patch been on the mailing list for more then 10 minutes because then it has a chance to be reviews by more people and and we will not get into a deal where maintainers make each one a favor of reviewing and pushing each other patches without anybody else get a chance to look at them.
> 
> Take a look at most other patches. You'll sadly see that most of them (though obviously not all) sat on the mailing list or PR for anything between 10 minutes and 10 days - and then got reviewed by one person and committed by the same person. In 75% of the cases this one person was Avi. So you are suggesting a great mechanism, but it's not how we've been doing things.

I do not see PRs that committed in 10 minutes (unless it is a typo) without anyone has a chance to reviews them except a committer (who is not the code maintainer). You create a bad precedent and try to justify it by something bad you allegedly saw happening before.


> 
> > And of course you need the code owner to review the code before you pushing it as a minimum.
> 
> Maybe, although this hasn't been the "minimum" until now.

It pretty much has been. By ignoring it you just delude it more since later you will point to this case to justify more of it.

> 
> > To have more chance to not introduce bugs you need to get an opportunity for the code to be reviewed by more people and even then they will be introduced.
> 
> There's always a tradeoff here between waiting for multiple reviews and quick commit, and it depends (I think) on the "risk" of the patch. If I don't feel comfortable in my understanding of the code, or feel that the person I'm reviewing doesn't know what he's doing - I'll usually wait for more reviewers. If I think the submitter of the code knew what he was doing and the patch doesn't have hard-to-understand algorithms and it's just some straightforward cleanup, I might think it doesn't need more reviews. This is what I did in this case.

Just ask the code owner to review. If he does not respond in a day or so and is not on vacation commit. The cleanup will wait for one day more. Not a tragedy to say the least.

> 
> BTW, you seem to be missing the fact that I _did_ review this code. It's not like I just took random junk code sent by a first-time contributer and put it in without even looking at it. It was code from a really good developer, and reviewed by me.

Basically what I get from your response is that you will commit any code from Avi for the code you do not maintain  unless it has absolutely obvious error in it. I do not thing this is a good judgement.

**Comment by nyh** [comment]
> I do not see PRs that committed in 10 minutes (unless it is a typo) without anyone has a chance to reviews them except a committer (who is not the code maintainer). You create a bad precedent and try to justify it by something bad you allegedly saw happening before.

The 10 minutes was just a fluke - it was past midnight (so most people were not working), I saw this new patch come in, and used LIFO order instead of FIFO order for my review.

But what I'm saying is that many other patches - especially simple code cleanups like this one - stay unreviewed for a few days before some committer reviews-and-commits-them in one pass. In the vast majority of the cases this reviewer-and-committer is Avi. Do you think Avi doesn't deserve the same courtesy and committers should be slower to commit his patches?

> > > And of course you need the code owner to review the code before you pushing it as a minimum.
> > Maybe, although this hasn't been the "minimum" until now.
> It pretty much has been. By ignoring it you just delude it more since later you will point to this case to justify more of it.

That has not been my observation - but I have to admit I did not try to gather statistics (and am not planning to do so now).

> Just ask the code owner to review. If he does not respond in a day or so and is not on vacation commit. The cleanup will wait for one day more. Not a tragedy to say the least.

Ok. A lot of times this ("waiting one day") happens by default - because the committers are slow so the code owners have more than a day to review. But a lot of times, the code owners don't use that more-of-the day, and one of the committers ends up both reviewing and committing.

> > BTW, you seem to be missing the fact that I _did_ review this code. It's not like I just took random junk code sent by a first-time contributer and put it in without even looking at it. It was code from a really good developer, and reviewed by me.
> 
> Basically what I get from your response is that you will commit any code from Avi for the code you do not maintain unless it has absolutely obvious error in it. I do not thing this is a good judgement.

I don't know where you got this. I said I did review the code. I never said I will commit **any** code from Avi. 
And I still don't think that this patch was as bad as you make it out to be. I don't think it's so terrible to go back and fix more of these get_local_storage_proxy() - especially since Avi explicitly said that he was going to do that. I still think that this patch was commit-worthy, even if it could have been better.

**Comment by gleb-cloudius** [comment]
> > I do not see PRs that committed in 10 minutes (unless it is a typo) without anyone has a chance to reviews them except a committer (who is not the code maintainer). You create a bad precedent and try to justify it by something bad you allegedly saw happening before.
> 
> The 10 minutes was just a fluke - it was past midnight (so most people were not working), I saw this new patch come in, and used LIFO order instead of FIFO order for my review.

Good to know it was just a fluke.

> 
> But what I'm saying is that many other patches - especially simple code cleanups like this one - stay unreviewed for a few days before some committer reviews-and-commits-them in one pass. In the vast majority of the cases this reviewer-and-committer is Avi. Do you think Avi doesn't deserve the same courtesy and committers should be slower to commit his patches?
> 
> > > > And of course you need the code owner to review the code before you pushing it as a minimum.
> > > > Maybe, although this hasn't been the "minimum" until now.
> > > > It pretty much has been. By ignoring it you just delude it more since later you will point to this case to justify more of it.
> 
> That has not been my observation - but I have to admit I did not try to gather statistics (and am not planning to do so now).

May be that because maintainers tend to commit each others code with relaxed requirements?
 

> 
> > Just ask the code owner to review. If he does not respond in a day or so and is not on vacation commit. The cleanup will wait for one day more. Not a tragedy to say the least.
> 
> Ok. A lot of times this ("waiting one day") happens by default - because the committers are slow so the code owners have more than a day to review. But a lot of times, the code owners don't use that more-of-the day, and one of the committers ends up both reviewing and committing.

Then a code owner cannot complain (unless vacation).

> 
> > > BTW, you seem to be missing the fact that I _did_ review this code. It's not like I just took random junk code sent by a first-time contributer and put it in without even looking at it. It was code from a really good developer, and reviewed by me.
> > 
> > 
> > Basically what I get from your response is that you will commit any code from Avi for the code you do not maintain unless it has absolutely obvious error in it. I do not thing this is a good judgement.
> 
> I don't know where you got this. 

Things like "It was code from a really good developer" hint to it.

> I said I did review the code.

You are not this code maintainer. Neither is the contributor. It does not mean you cannot review of course. But it should not be enough.


> I never said I will commit **any** code from Avi. And I still don't think that this patch was as bad as you make it out to be. I don't think it's so terrible to go back and fix more of these get_local_storage_proxy() - especially since Avi explicitly said that he was going to do that. I still think that this patch was commit-worthy, even if it could have been better.

I do not say the patch terrible.  What I am saying that procedure should be followed no matter who submitter is (unless the patch is no more then a typo fix).

**Comment by nyh** [comment]
> I do not say the patch terrible. What I am saying that procedure should be followed no matter who submitter is (unless the patch is no more then a typo fix).

Please document the procedure which you insist must be followed, and then we can review this procedure you are proposing.
Until now, there  hasn't been any ironclad procedure, just common sense plus a handful of rules like "don't commit your own code". 

Keep in mind that having a patch both reviewed and committed by one person (who is not the author of the patch) happened very frequently in the past. Did you ever complain when Avi committed a patch after having reviewing it himself? Will the procedure you propose prevent Avi from ever single-handedly reviewing a patch and then committing it, or require him to wait a day (or a work day, or a week, or?) until he does so?

**Comment by gleb-cloudius** [comment]
> > I do not say the patch terrible. What I am saying that procedure should be followed no matter who submitter is (unless the patch is no more then a typo fix).
> 
> Please document the procedure which you insist must be followed, and then we can review this procedure you are proposing. Until now, there hasn't been any ironclad procedure, just common sense plus a handful of rules like "don't commit your own code".

Ask whoever familiar with the code to review. The rule is old as OS.

> 
> Keep in mind that having a patch both reviewed and committed by one person (who is not the author of the patch) happened very frequently in the past. Did you ever complain when Avi committed a patch after having reviewing it himself? Will the procedure you propose prevent Avi from ever single-handedly reviewing a patch and then committing it, or require him to wait a day (or a work day, or a week, or?) until he does so?

If he will do to a code I am responsible for without giving me even a little chance to do a review I will complain. My experience with him is different though - he asks for reviews.

**Comment by nyh** [comment]
> If he will do to a code I am responsible for without giving me even a little chance to do a review I will complain. My experience with him is different though - he asks for reviews.

This is my last mail in this thread. If you want to continue this conversation, please do what I suggested: Write down the rules you insist that maintainers must abide with, and then we can discuss them and all maintainers will abide by them - for better and for worse.

As for your last statement, I looked at the last commits, and randomly found "gms: gossiper: coroutinize a few small functions." PavelS sent the 4-patch series on the 13th, nobody reviewed it, so on the 15th he asked again for review - at which point Avi reviewed and committed it himself. Nobody waited or specifically reminded the "owner" of this code, and Avi thought he is capable enough to review this patch himself. It seems the only difference between that and my case was that I didn't wait for two days with my review. Sorry for being so efficient, next time I'll try to be less efficient ;-)

**Comment by gleb-cloudius** [comment]
> > If he will do to a code I am responsible for without giving me even a little chance to do a review I will complain. My experience with him is different though - he asks for reviews.
> 
> This is my last mail in this thread. If you want to continue this conversation, please do what I suggested: Write down the rules you insist that maintainers must abide with, and then we can discuss them and all maintainers will abide by them - for better and for worse.
> 
> As for your last statement, I looked at the last commits, and randomly found "gms: gossiper: coroutinize a few small functions." PavelS sent the 4-patch series on the 13th, nobody reviewed it, so on the 15th he asked again for review - at which point Avi reviewed and committed it himself. Nobody waited or specifically reminded the "owner" of this code, and Avi thought he is capable enough to review this patch himself. It seems the only difference between that and my case was that I didn't wait for two days with my review. Sorry for being so efficient, next time I'll try to be less efficient ;-)

This is not efficiency this is rushing things out without proper procedure. If Asias did not review trivial patchs for two days it is OK to commit then without his review. And if you would have committed this patch tomorrow for  instance we would not have had this conversation either, so good job missing the point.

**Comment by gleb-cloudius** [comment]
And sometimes it is not even about patch or review quality.  Both can be perfect but it still worth to wait for other consideration. Like the patch with data dictionary that was luckily not committed immediately because of terminology dispute, but if it would have been it would have created a huge churn for 60 patches schema series.

**Pull request closed without merging**
